clang/lib/Sema/SemaBounds.cpp

```diff
@@ -3162,6 +3233,136 @@ namespace {
+    // SubExprGs stores, for each subexpression Si of e, a set Gi
+    // of expressions that produce the same value as Si.
+    //
+    // Val is an optional expression that may be contained in the updated G.
+    // If Val is not provided, e is used instead.
+    void UpdateG(Expr *E, ExprEqualMapTy SubExprGs,
+                 EqualExprTy &G, Expr *Val = nullptr) {
+      G.clear();
+
+      if (!Val) Val = E;
+
+      // If Val is a call expression, G does not contain Val.
+      if (isa<CallExpr>(Val)) {
+      }
+
+      // If Val is a non-modifying expression, G contains Val.
+      else if (CheckIsNonModifying(Val))
+        G.push_back(Val);
+
+      // If Val is a modifying expression, use the Gi for the subexpressions
+      // to try to construct a non-modifying expression Val' that
+      // produces the same value as Val.
+      else {
+        Expr *ValPrime = nullptr;
+        for (llvm::detail::DenseMapPair<Expr *, EqualExprTy> Pair : SubExprGs) {
+          Expr *Si = Pair.first;
+          // For any modifying subexpression Si of e,
+          // try to set Val' to a nonmodifying expression from Gi.
+          if (!CheckIsNonModifying(Si)) {
+            EqualExprTy Gi = Pair.second;
+            for (auto I = Gi.begin(); I != Gi.end(); ++I) {
+              Expr *Ei = *I;
+              if (CheckIsNonModifying(Ei)) {
+                ValPrime = Ei;
+                break;
+              }
+            }
+          }
+        }
+
+        if (ValPrime)
+          G.push_back(ValPrime);
+      }
+
+      // If Val introduces a temporary to hold the value produced by e,
+      // add the value of the temporary to G.
+      if (CHKCBindTemporaryExpr *Temp = GetTempBinding(Val))
+        G.push_back(CreateTemporaryUse(Temp));
+    }
+
+    // If E appears in a set F in EQ, GetEqualExprSetContainingExpr
+    // returns F.  Otherwise, it returns an empty set.
+    EqualExprTy GetEqualExprSetContainingExpr(Expr *E, EquivExprSets EQ) {
+      for (auto OuterList = EQ.begin(); OuterList != EQ.end(); ++OuterList) {
+        EqualExprTy F = *OuterList;
+        for (auto InnerList = F.begin(); InnerList != F.end(); ++InnerList) {
+          Expr *E1 = *InnerList;
+          if (EqualValue(S.Context, E, E1, &EQ))
+            return F;
+        }
+      }
+      return { };
+    }
+
+    // Returns true if the lvalue expression e reads memory via a pointer.
+    bool ReadsMemoryViaPointer(Expr *E) {
+      E = E->IgnoreParens();
+
+      if (!E->isLValue())
+        return false;
+
+      switch (E->getStmtClass()) {
+        case Expr::UnaryOperatorClass: {
+          UnaryOperator *UO = cast<UnaryOperator>(E);
+          // *e reads memory via a pointer.
+          return UO->getOpcode() == UnaryOperatorKind::UO_Deref;
+        }
+        // e1[e2] is a synonym for *(e1 + e2), which reads memory via a pointer.
+        case Expr::ArraySubscriptExprClass:
+          return true;
+        // e1.e2 and e1->e2 read memory via a pointer.
```

Comment:
I think the proper syntax is e1.f or e1->f.   In the first case, e1.f reads memory via a pointer only of e1 reads memory via a pointer.